quickwit/quickwit-ingest-api/src/ingest_api_service.rs

```diff
@@ -90,44 +96,57 @@ impl IngestApiService {
             queues,
             memory_limit,
             disk_limit,
+            capacity: Arc::new(AtomicIsize::new(memory_limit as isize)),
+            capacity_notify: Arc::new(Notify::new()),
         })
     }
 
-    async fn ingest(
-        &mut self,
-        request: IngestRequest,
-        ctx: &ActorContext<Self>,
-    ) -> crate::Result<IngestResponse> {
+    async fn ingest_inner(&mut self, request: IngestRequest) -> crate::Result<IngestResponse> {
         // Check all indexes exist assuming existing queues always have a corresponding index.
-        let first_non_existing_queue_opt = request
+        for doc_batch in &request.doc_batches {
+            if !self.queues.queue_exists(&doc_batch.index_id).await {
+                return Err(IngestServiceError::IndexNotFound {
+                    index_id: doc_batch.index_id.clone(),
+                });
+            }
+        }
+        let payload_size = request
             .doc_batches
             .iter()
-            .map(|batch| batch.index_id.as_str())
-            .find(|index_id| !self.queues.queue_exists(index_id));
+            .map(|doc_batch| doc_batch.doc_lens.iter().sum::<u64>())
+            .sum::<u64>() as isize;
+        loop {
+            let current_capacity = self.capacity.load(Ordering::Acquire);
+            if current_capacity > 0 {
+                let new_capacity = current_capacity - payload_size;
 
-        if let Some(index_id) = first_non_existing_queue_opt {
-            return Err(IngestServiceError::IndexNotFound {
-                index_id: index_id.to_string(),
-            });
-        }
-
-        let (memory, disk) = self.queues.ressource_usage();
-        if memory > self.memory_limit {
-            info!("Ingestion rejected due to memory limits");
-            return Err(IngestServiceError::RateLimited);
-        }
-        if disk > self.disk_limit {
-            info!("Ingestion rejected due to disk limits");
-            return Err(IngestServiceError::RateLimited);
+                if new_capacity >= 0 {
+                    if self
+                        .capacity
+                        .compare_exchange(
+                            current_capacity,
+                            new_capacity,
+                            Ordering::AcqRel,
+                            Ordering::Acquire,
+                        )
+                        .is_ok()
+                    {
+                        if new_capacity > 0 {
+                            self.capacity_notify.notify_one();
+                        }
+                        break;
+                    }
+                }
+            }
+            self.capacity_notify.notified().await;
```

Comment:
If truncate is never called, this waits forever :/

Reply:
Note to self:
- wrap with timeout and return 429
- count the number of pending clients and return if above > some threshold?